quinn/src/recv_stream.rs

```diff
@@ -284,6 +284,54 @@ impl RecvStream {
         self.stream
     }
 
+    /// Completes when the stream is reset by the peer or otherwise closed
+    ///
+    /// **Does not** actively cause the stream to become reset. Contrast [`stop()`](Self::stop),
+    /// which notifies the peer that the stream will no longer be read and should be reset.
+    ///
+    /// Yields `Some` with the reset error code when the stream is reset by the peer. Yields `None`
+    /// when the stream was previously [`stop()`](Self::stop)ed, or when the stream was
+    /// [`finish()`](crate::SendStream::finish)ed by the peer and all data has been received, after
+    /// which it is no longer meaningful for the stream to be reset.
+    ///
+    /// Notification of a reset is delivered exactly once. After a reset error code has been
+    /// delivered by *any* means, including calls to `reset()` or any operation that returns
+    /// [`ReadError::Reset`], future calls to `reset()` will return `Ok(None)`, and future calls to
+    /// methods that can return [`ReadError`] will return [`ReadError::ClosedStream`].
+    ///
+    /// This operation is cancel-safe.
+    pub async fn reset(&mut self) -> Result<Option<VarInt>, ResetError> {
+        poll_fn(|cx| {
+            let mut conn = self.conn.state.lock("RecvStream::reset");
+            if self.is_0rtt && conn.check_0rtt().is_err() {
+                return Poll::Ready(Err(ResetError::ZeroRttRejected));
+            }
+
+            if let Some(code) = self.reset {
+                return Poll::Ready(Ok(Some(code)));
+            }
```

Comment:
You documented that notification of a reset would be delivered exactly once, but that doesn't seem to match this code?

Reply:
This was stale from a previous revision. Corrected, and added a similar warning to the `proto` layer where it *does* hold.